takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/compile/jdt/ProcessingEnvImpl.java

```diff
@@ -1,36 +1,131 @@
 
 import java.util.Collections;
+import java.util.LinkedHashSet;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
 
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.PackageElement;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
 import javax.tools.StandardJavaFileManager;
 
+import org.eclipse.jdt.core.compiler.CharOperation;
 import org.eclipse.jdt.internal.compiler.Compiler;
 import org.eclipse.jdt.internal.compiler.apt.dispatch.BaseProcessingEnvImpl;
+import org.eclipse.jdt.internal.compiler.apt.model.ElementsImpl;
+import org.eclipse.jdt.internal.compiler.apt.model.Factory;
+import org.eclipse.jdt.internal.compiler.lookup.Binding;
+import org.eclipse.jdt.internal.compiler.lookup.ImportBinding;
+import org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment;
+import org.eclipse.jdt.internal.compiler.lookup.PackageBinding;
+import org.eclipse.jdt.internal.compiler.lookup.ReferenceBinding;
+import org.eclipse.jdt.internal.compiler.lookup.TypeBinding;
 
 import io.takari.maven.plugins.compile.AbstractCompileMojo.Proc;
 import io.takari.maven.plugins.compile.CompilerBuildContext;
 
 // TODO reconcile with BatchProcessingEnvImpl
 class ProcessingEnvImpl extends BaseProcessingEnvImpl {
 
+  private Set<Consumer<String>> referencedTypeObservers = new LinkedHashSet<>();
+
+  // TODO this shadows private member of the superclass, not pretty
+  private final Factory _factory;
+
   public ProcessingEnvImpl(CompilerBuildContext context, StandardJavaFileManager fileManager, Map<String, String> processorOptions, Compiler compiler, CompilerJdt incrementalCompiler, Proc proc) {
     this._filer = new FilerImpl(context, fileManager, incrementalCompiler, this, proc);
     this._messager = new MessagerImpl(context, this);
     this._processorOptions = processorOptions != null ? processorOptions : Collections.<String, String>emptyMap();
     this._compiler = compiler;
+
+    this._elementUtils = new ElementsImpl(this) {
+      @Override
+      public TypeElement getTypeElement(CharSequence name) {
+        observeType(name.toString());
+        return super.getTypeElement(name);
+      }
+
+      @Override
+      public PackageElement getPackageElement(CharSequence name) {
+        observeType(name.toString());
+        return super.getPackageElement(name);
+      }
+    };
+
+    this._factory = new Factory(this) {
+      private void observeType(Binding binding) {
+        if (binding instanceof ImportBinding) {
+          observeType(((ImportBinding) binding).compoundName);
+        } else if (binding instanceof PackageBinding) {
+          observeType(((PackageBinding) binding).compoundName);
+        } else if (binding instanceof TypeBinding) {
+          observeType((TypeBinding) binding);
+        }
+        // note that variable reference does not imply reference to variable type
```

Comment:
Without the type references however how would the processor get notified of a change to the method or variable?   If you reference part of it, isn't it a reference to all of it.   Generic parameters and all that could effect return types etc.    Seems unsafe to throw away these type references.